File: game.fs
\ Wordle game

0 value secret      ( the secret word we are trying to guess )

: score-guess ( guess -- score )  secret swap score ;

: solved ( score -- f ) [s] ggggg = ;
: failed ( -- f )       guesses 5 > ; ( assuming not solved )

\ Maintain greens for solver convenience, letter or '-'
create greens len allot
: #greens ( -- n )  len  greens len bounds do  i c@ '-' = +  loop ;
: +greens ( guess score -- )  swap ww swap
    len 0 do  3 /mod swap green = if  over i + c@  i greens + c!  then
    loop 2drop ;

\ Like the wordle-game's keyboard, record what we know about each letter
\ 0=unknown, 1=grey, 2=yellow, 3=green (k = color+1)
create kb 32 allot
: >kb ( c -- a )  31 and kb + ;
: +kb ( k c -- )  >kb 2dup c@ > if c! else 2drop then ;
: +keyboard ( guess score -- ) \ update kb with latest guess & score
    swap ww len bounds do  3 /mod swap ( color ) 1+ i c@ +kb  loop drop ;

: .game ." secret: " secret w. ." guesses: " guesses .
        ." greens: " greens len type space ;

\ Submit guess to the game, update game state and return the score
: submit ( guess score -- )  2dup +history  2dup +greens  +keyboard ;
: guess  ( guess -- score )  dup score-guess  tuck submit ;

\ Initialize everything except the secret (reset game)
: init-game ( -- )  0 to guesses  greens len '-' fill  kb 32 erase ;

\ Initialize a new game and pick a random secret word
: new-game ( -- )  init-game  #hidden random to secret ;



( ===== TESTS ===== )
testing +greens
init-game
t{ w raise s gy-g- +greens  s" r--s-" greens len compare -> 0 }t
t{ w smack s -g--- +greens  s" rm-s-" greens len compare -> 0 }t
t{ w pound s ----- +greens  s" rm-s-" greens len compare -> 0 }t

testing #greens
t{ s" -----" greens swap move   #greens -> 0 }t
t{ s" -i-d-" greens swap move   #greens -> 2 }t
t{ s" windy" greens swap move   #greens -> 5 }t

File: rust/src/game.rs
// Wordle game

use crate::words::{Word, ww, find};
use crate::score::{Score, score, GREEN};

pub const GUESSES: usize = 6;     // max guesses

#[derive(Default)]
pub struct Game {
    secret: Word,           // the secret word
    guesses: usize,         // number of guesses, 0-6
    history: [(Word, Score); GUESSES],   // history of guesses
    used: [u8; 32],         // letter usage, 0=?, 1=grey, 2=yellow, 3=green
    greens: [char; 5],      // green letter or '-' if not known
}

impl Game {
    pub fn new() -> Game {
        let secret: Word = 1234; // todo: random
        Self::with_secret(secret)
    }

    pub fn with_secret(secret: Word) -> Game {
        Game {
            secret: secret,
            guesses: 0,
            history: [(0,0); GUESSES],
            used: [0; 32],
            greens: ['-'; 5],
        }
    }

    pub fn secret(&mut self, w: Word) {
        self.secret = w;
    }

    pub fn reset(&mut self) {
        self.guesses = 0;
        self.used.fill(0);
        self.greens.fill('-');
    }

    pub fn guesses(&self) -> usize {
        self.guesses
    }

    pub fn latest(&self) -> Option<(Word, Score)> {
        if self.guesses > 0 {
            Some(self.history[self.guesses-1])
        } else {
            None
        }
    }

    pub fn greens(&self) -> i32 {
        let mut greens = 0;
        for c in self.greens {
            if c != '-' {
                greens += 1;
            }
        }
        greens
    }

    pub fn answer(&self) -> Word {
        let s: String = self.greens.iter().collect();
        find(s.as_str()).unwrap()
    }

    pub fn submit(&mut self, guess: Word) -> bool {
        if self.guesses >= GUESSES {
            panic!("no more guesses");
        }

        let score = score(guess, self.secret);
        let solved = score == 242;

        self.history[self.guesses] = (guess, score);
        self.guesses += 1;

        // update used and greens
        let g: Vec<char> = ww(guess).chars().collect();
        let mut mscore = score;
        for i in 0..5 {
            let c = g[i];
            let s = mscore % 3;
            mscore /= 3;

            let l = c as usize & 31;
            if s >= self.used[l] {
                self.used[l] = s + 1;
            }

            if s == GREEN {
                self.greens[i] = c;
            }
        }

        solved
    }
}
